src/compiler/checker.ts

```diff
@@ -1040,19 +1061,28 @@ module ts {
             return result || emptyArray;
         }
 
-        function setObjectTypeMembers(type: ObjectType, members: SymbolTable, callSignatures: Signature[], constructSignatures: Signature[], stringIndexType: Type, numberIndexType: Type): ResolvedType {
+        function setObjectTypeMembers(type: ObjectType, members: SymbolTable, callSignatures: Signature[], constructSignatures: Signature[], stringIndex: IndexType, numberIndex: IndexType): ResolvedType {
             (<ResolvedType>type).members = members;
             (<ResolvedType>type).properties = getNamedMembers(members);
             (<ResolvedType>type).callSignatures = callSignatures;
             (<ResolvedType>type).constructSignatures = constructSignatures;
-            if (stringIndexType) (<ResolvedType>type).stringIndexType = stringIndexType;
-            if (numberIndexType) (<ResolvedType>type).numberIndexType = numberIndexType;
+            if (stringIndex) (<ResolvedType>type).stringIndex = stringIndex;
+            if (numberIndex) (<ResolvedType>type).numberIndex = numberIndex;
+
+            if (stringIndex && numberIndex) {

```

Comment:
It's mainly to keep around metadata about where the index was inherited from:
https://github.com/Microsoft/TypeScript/compare/master...jbondc:issue-2491#diff-c3ed224e4daa84352f7f1abcd23e8ccaR10055

Either both indexes (YES) are from the class 'foo' extends a, b {} or both indexes (INHERITED) are from 'e' in class d extends e {}

Ideally I'd like to remove it, couldn't find a general solution to keeping around the nodes we could want to reports errors on.

Specifically this test here:
https://github.com/Microsoft/TypeScript/compare/master...jbondc:issue-2491#diff-7169c6ef1ec96c9d75b987be9f53f6c7L73

Ideally the errors should be reported on the node which matches the members:
e.g. on 'B' for where Property '1' of type 'Z'